concurrent_map.go

```diff
@@ -5,75 +5,100 @@ import (
 	"sync"
 )
 
-var SHARD_COUNT = 32
+// ShardCount represent the number of shard the map is devided into.
+var ShardCount = 100
 
-// A "thread" safe map of type string:Anything.
-// To avoid lock bottlenecks this map is dived to several (SHARD_COUNT) map shards.
+// ConcurrentMap is a "thread" safe map of type string:Anything.
+// To avoid lock bottlenecks this map is dived to several (ShardCount) map shards.
 type ConcurrentMap []*ConcurrentMapShared
 
-// A "thread" safe string to anything map.
+// ConcurrentMapShared is a "thread" safe string to anything map.
 type ConcurrentMapShared struct {
+	mapKeys      []string
+	maxSize      int
 	items        map[string]interface{}
 	sync.RWMutex // Read Write mutex, guards access to internal map.
 }
 
-// Creates a new concurrent map.
-func New() ConcurrentMap {
-	m := make(ConcurrentMap, SHARD_COUNT)
-	for i := 0; i < SHARD_COUNT; i++ {
-		m[i] = &ConcurrentMapShared{items: make(map[string]interface{})}
+// New creates a new concurrent map.
+func New(maxSize int) ConcurrentMap {
+	m := make(ConcurrentMap, ShardCount)
+	shardKeySize := maxSize / ShardCount
+	for i := 0; i < ShardCount; i++ {
+		m[i] = &ConcurrentMapShared{
+			maxSize: shardKeySize,
+			items:   make(map[string]interface{}),
+		}
 	}
 	return m
 }
 
-// GetShard returns shard under given key
+// GetShard returns shard under given key.
 func (m ConcurrentMap) GetShard(key string) *ConcurrentMapShared {
-	return m[uint(fnv32(key))%uint(SHARD_COUNT)]
+	return m[uint(fnv32(key))%uint(ShardCount)]
 }
 
 func (m ConcurrentMap) MSet(data map[string]interface{}) {
 	for key, value := range data {
+		if m.isEvictionOccurred(key) {
+			m.removeFirstElement(key)
+		}
 		shard := m.GetShard(key)
 		shard.Lock()
 		shard.items[key] = value
+		shard.mapKeys = append(shard.mapKeys, key)
 		shard.Unlock()
 	}
 }
 
-// Sets the given value under the specified key.
+// Set sets the given value under the specified key.
 func (m ConcurrentMap) Set(key string, value interface{}) {
 	// Get map shard.
 	shard := m.GetShard(key)
+	if m.isEvictionOccurred(key) {
+		m.removeFirstElement(key)
+	}
 	shard.Lock()
 	shard.items[key] = value
+	shard.mapKeys = append(shard.mapKeys, key)
 	shard.Unlock()
 }
 
-// Callback to return new element to be inserted into the map
+// UpsertCb is callback function to return new element to be inserted into the map.
 // It is called while lock is held, therefore it MUST NOT
 // try to access other keys in same map, as it can lead to deadlock since
 // Go sync.RWLock is not reentrant
 type UpsertCb func(exist bool, valueInMap interface{}, newValue interface{}) interface{}
 
-// Insert or Update - updates existing element or inserts a new one using UpsertCb
+// Upsert (Insert or Update) - updates existing element or inserts a new one using UpsertCb.
 func (m ConcurrentMap) Upsert(key string, value interface{}, cb UpsertCb) (res interface{}) {
 	shard := m.GetShard(key)
+
+	if m.isEvictionOccurred(key) {
```

Comment:
The `Upsert` function is not an interface method, which means it is not used in our project, however your statement is right! I have adapted the code accordingly. Also in case of eviction the key is removed from the map (this was missing).